### DISCUSSION TITLE
OBS Browser/CEF Plans, Progress & Timeline

### DISCUSSION THREAD

**Discussion by WizardCM** [obsproject/obs-studio#3853] [Announcements]
This page is designed to show current progress/plans when it comes to improvements, fixes, and feature parity when it comes to the Browser Source, Custom Browser Docks & Service Integration. Questions/discussion welcome below.

Previously I was posting status updates in the #development channel on Discord but felt they'd be very easily lost.

Note that actual implementation details, PR reviews & bugs should be posted on the respective Issues or PRs.

-------

Edit February 20, 2025: OBS 31 was released with Chromium 127. Changes on hold for the moment due to Chrome Runtime, but shouldn't be near as painful to upgrade as last time as we're not waiting on outside parties. [See here](https://github.com/obsproject/obs-studio/discussions/3853#discussioncomment-12272738) for details.

Edit June 9, 2024: Big update in regards to future CEF upgrades, including successful testing of Chromium 126 across Windows, macOS, Linux. [See here](https://github.com/obsproject/obs-studio/discussions/3853#discussioncomment-9718414) for the full writeup.

Edit September 1, 2022: OBS 28.0 was released with Chromium 103. Changes are on hold due to changes in Chromium that make our existing patches for shared textures incompatible.

Edit November 28, 2021: [See here](https://github.com/obsproject/obs-studio/discussions/3853#discussioncomment-1710556) regarding OBS 27.2.

Edit April 18, 2021: [See here](https://github.com/obsproject/obs-studio/discussions/3853#discussioncomment-625248) regarding 26.1.2 and 27.0.

-------

#### Version
* Windows: CEF 3770 (Chromium 75)
* macOS: CEF 4183 (Chromium 85)
* Linux: CEF 4280 (Chromium 87)

#### Feature parity
* Custom Browser Docks: Windows, macOS, Linux (except Wayland)
* Service Integration: Windows, macOS, Linux (except Wayland)

-------

## Major plans for the browser source

**Note:** For each of the features described below, every item has to be done for a feature to be ready. Estimates are hopeful, and will likely change. There is no actual, concrete timeline.

* [x] **Browser Panels & Service integration on macOS** https://github.com/obsproject/obs-browser/pull/197

  * Estimated: OBS 26.2 / 27

  * Note: It's possible that these two capabilities will be added in separate versions

* [x] **Browser Panels & Service integration on Linux** https://github.com/obsproject/obs-browser/pull/254

  * Estimated: OBS 27

  * Note: It's possible that these two capabilities will be added in separate versions

* [x] **Update to Chromium 95 / CEF 4638 on Windows, macOS and Linux**

  * Estimated: OBS 27.2

### Update to Chromium 95 / CEF 4638

The main purpose here is to use a more recent version of Chromium.

* [x] Shared textures overhaul (& hardware acceleration) on Windows **complex**

* * [x] Update libobs & obs-browser to support new shared texture handles

* * [x] Performance testing

* [x] Test disabling hardware acceleration and opening 5+ browser sources

* [x] Test downgrading from newer CEF to older build (when downgrading OBS itself) - mostly cookies

* [x] Upload new CEF builds to Fastly for both macOS and Windows

-------

### Have any questions? Post them below!

**Comment by WizardCM** [comment]
Status Update:

### As of OBS 26.1.2:

* Successfully updated to CEF 4183 (Chrome 85) on **macOS** - a little earlier than expected

### As of OBS 27.0:

* Successfully updated to CEF 4280 (Chrome 87) on **Linux**
* Successfully added Service Integration / Custom Browser docks on **macOS** and **Linux**

### And then there's Windows..

Let me be up front: we **absolutely** want to update CEF **as soon as possible**. Unfortunately, we were unable to get it updated in time for OBS 27.0 due to complications with the shared texture API provided in the unmerged CEF PR.

The original PR for CEF was designed for Chrome 79, and was adapted to work with Chrome 85. There are builds made for Chrome 87, but they don't seem to work, and we were unable to get it to work with Chrome 89 either. Likely something changed on the Chrome side.

**We plan to revisit the CEF update after 27.0 is out, to try and get it working for 27.1.**

**Reply by JillKatze** [reply]
Of course I stumbled across all of this _after_ trying to do a local build with a new CEF drop to see if it fixed obsproject/obs-browser#237 and having it explode spectacularly 💀

As I mentioned in that issue, I believe the audio popping/stuttering/crackling that sometimes occurs when using "Control audio via OBS" on browser sources may have been due to a browser bug that was fixed recently (see [Chromium bug 1130375](https://bugs.chromium.org/p/chromium/issues/detail?id=1130375)). I wanted to make sure this particular commit was on the radar of whoever ends up doing the work to get a new CEF drop working in Windows, so that hopefully a version that contains that fix gets pulled in.

I'm no expert in how Chromium branches and versioning work, but from the tags it seems like that was merged into 89 (4389) and 90 (4430), so if this issue also affects Mac and Linux (unsure, don't have the ability to test those atm), then they would also benefit from bumping up again to a version with this change, if it actually fixes the issue in the browser source.

My apologies if this is potentially making more work for someone, but it would be disappointing to see work get put in to update CEF in the next release but to a version just old enough to miss this fix I really want. Thanks for your time.

**Reply by WizardCM** [reply]
Thanks for reaching out and documenting your findings. We will certainly keep this issue in mind when updating CEF on Windows.

**Comment by FaeyUmbrea** [comment]
It would generally be amazing if the CEF version could be updated more frequently than once every 2 years or so. There is a bunch of JS language features missing due to the large gap in between update. One of the bigger features missing in 75 is the "?." operator, also known as Optional Chaining which was added in Chromium 80, this creates dependency on using babel.js to create larger, more unoptimized JS files if one wants their site to work in OBS.

**Reply by dodgepong** [reply]
I don't think anyone objects to the desire to update more frequently. The point, here, is that the reason CEF hasn't been updated isn't due to laziness or a lax update schedule, but because the newer versions of CEF literally lack features that OBS absolutely requires to run effectively. Do you really want to sacrifice hardware acceleration in order to get option chaining? I think most people would say no.

**Comment by DerValiser** [comment]
Please update at least to version 88 of Chromium.  The latest version is 95, by the way.

The planned Version 85 still does not support clip-path path() which makes SVG animations nearly impossible in OBS.

https://caniuse.com/mdn-css_properties_clip-path_path

See this example, which is not supported in obs: 
https://wiki.selfhtml.org/extensions/Selfhtml/example.php/Beispiel:CSS3_clip-path-3.html

**Reply by Fenrirthviti** [reply]
Please read the thread here for the listed reasons on why it's not that simple.

**Reply by dodgepong** [reply]
I feel like you didn't even read anything that has been written so far.

**Reply by RytoEX** [reply]
As explained [here](https://github.com/obsproject/obs-studio/discussions/3853#discussioncomment-625248):
> Let me be up front: we absolutely want to update CEF as soon as possible. Unfortunately, we were unable to get it updated in time for OBS 27.0 due to complications with the shared texture API provided in the unmerged CEF PR.
> 
> The original PR for CEF was designed for Chrome 79, and was adapted to work with Chrome 85. There are builds made for Chrome 87, but they don't seem to work, and we were unable to get it to work with Chrome 89 either. Likely something changed on the Chrome side.

and [here](https://github.com/obsproject/obs-studio/discussions/3853#discussioncomment-822027):
> the reason CEF hasn't been updated isn't due to laziness or a lax update schedule, but because the newer versions of CEF literally lack features that OBS absolutely requires to run effectively.

We understand the desire to have new CSS and JS features available.  However, this isn't a simple matter of us just downloading a new version.  If we were to simply ship with a new version of CEF without Shared Texture support, CPU usage of browser sources would increase drastically and performance would decrease.

Thank you for your patience and understanding.

**Reply by DerValiser** [reply]
@RytoEX  Thank you for the kind explanation.

**Reply by DerValiser** [reply]
Although it would be interesting to know, which features should be missing in the latest CEF version.

**Reply by RytoEX** [reply]
> 
> 
> Although it would be interesting to know, which features should be missing in the latest CEF version.

Shared Textures, as mentioned above.  We're relying on an unmerged PR to CEF to provide Shared Textures functionality, and that PR cannot simply be merged into newer versions of CEF.  We have been unable to adapt it to newer versions of CEF thus far.  That is the main reason for the delay.

**Reply by DerValiser** [reply]
I assume you already asked the team there if it would be possible to merge this pull request? Maybe asking them, in case you are just waiting, would speed up the process. Seems that your life would be easier as well then. Good luck and thank you for the work.

**Reply by Fenrirthviti** [reply]
The answers to all of these questions are in the thread, I suggest taking the time at this point to actually read the discussion to avoid asking the same questions that have already been answered.

**Reply by FaeyUmbrea** [reply]
Actually if I could hijack this thread for a second. There is no link to the actual discussions on the side of CEF and I have lost it after looking it up in my original comment above this one. It would be amazing if you could provide those links or regular updates to this topic on how the project is going. What ever is easiest for you and the team.
From the outside the lack of new information makes this process look abandonned.

**Reply by Fenrirthviti** [reply]
Thanks, my mistake I had thought the links to the CEF PR (where a lot of the discussion took place) were linked here.

[CEF PR #285 (shared texture support)](https://bitbucket.org/chromiumembedded/cef/pull-requests/285)

**Reply by DerValiser** [reply]
> 
> 
> The answers to all of these questions are in the thread, I suggest taking the time at this point to actually read the discussion to avoid asking the same questions that have already been answered.

This is why you shall never rely on PRs. Simple developer rule.

**Reply by RytoEX** [reply]
> 
> 
> I assume you already asked the team there if it would be possible to merge this pull request? Maybe asking them, in case you are just waiting, would speed up the process. Seems that your life would be easier as well then. Good luck and thank you for the work.

The maintainer of CEF has already stated that they do not plan on merging the PR without the Chromium team implementing and maintaining some portion of the changes necessary to support the PR.  That is more or less where the conversation ended quite some time ago.  To my recollection, we have already expressed interest in working with other groups to maintain the PR so that we can all benefit from the performance benefits of the feature.

> 
> 
> > The answers to all of these questions are in the thread, I suggest taking the time at this point to actually read the discussion to avoid asking the same questions that have already been answered.
> 
> This is why you shall never rely on PRs. Simple developer rule.

If we did not utilize the PR in question, the browser source would simply not be nearly as performant as it is.  Asking us to simply do away with that would be unnecessarily inflicting performance penalties on every browser source instance which, in some cases, would be so detrimental it would become impossible to utilize them in a stable fashion.

If you have the time or resources to assist us in our efforts to get the PR working for a newer version of CEF, that would be welcome.

**Comment by pjtgrizzly** [comment]
Is Update to Chromium 85 / CEF 4183 on Windows still slated for ballpark of OBS 27.2? Just curious if this got pushed back further at this point as assuming with 27.1 just coming out and didn't seem to include this update that perhaps it's on the very near horizon?

**Reply by WizardCM** [reply]
Unfortunately there was not a lot of time to revisit this - however it's not looking promising. I will post an update as soon as I have something more concrete.

**Comment by pjtgrizzly** [comment]
Got it - thanks so much for all y'all do!

**Comment by WizardCM** [comment]
Status Update:

As some noticed, we did not manage to squeeze a CEF update into OBS 27.1.

However, I come bearing really, really great news:

### We've successfully built & tested CEF 4638 (Chromium 95) with all the bells & whistles we require. 🎉🎉

```
15:42:58.079: [obs-browser]: Version 2.16.3
15:42:58.079: [obs-browser]: CEF Version 95.7.9+g0d1bea3+chromium-95.0.4638.17
```

As mentioned before, Windows was a very big challenge. Thanks to the amazing efforts of @pkviet, @VodBox, @PatTheMav and the original author of the CEF PR that we use for shared textures, things are looking up.

As it stands, there are a couple of smaller blockers we still need to work on, but **I am hopeful that this will ship in OBS 27.2** (on all platforms). 🚀

PRs to keep an eye on:

* https://github.com/obsproject/obs-browser/pull/323
* https://github.com/obsproject/obs-studio/pull/5518

Changes that still need to be made/tested:

* [x] `file://` behaviours have changed dramatically in Chromium, so we will have to perform different workarounds on our side. I personally hope to look into this over the coming weeks
* [x] Testing still needs to be done on Linux

Other issues may come up as we test, and I will post this as we go.

**Reply by pjtgrizzly** [reply]
That is awesome - talk about brand spanking new version of Chromium. Looking forward - great work y'all!

**Reply by Trouffman** [reply]
This is promising! Thanks for the continued hard work on this!

**Reply by WizardCM** [reply]
I've made some minor updates to the above status update, mainly the addition of https://github.com/obsproject/obs-studio/pull/5518 and that I've done some testing of `file://` mode. I hope to post another proper status update soon.

**Comment by WizardCM** [comment]
Status Update:

### The CEF 95 update has been merged for OBS 27.2. 🎉🎉

All our tests on Windows, macOS and Linux have passed with flying colours. The code has now been merged, and we're all set to ship CEF 95.

We did run into small quicks that needed to be fixed along the way, but everything was easily solvable. CI successfully built with existing versions so if we do have to revert for any reason (which we likely won't have to), then those parts are safe too.

Next steps:

* [x] Upload CEF 95 to our Fastly mirror
* [x] Update CI with new links to CEF 95

I will post an update on this comment when the two above tasks are done.

**Reply by WizardCM** [reply]
* https://github.com/obsproject/obs-studio/pull/5648

The pull request with CEF 95 on CI is now up! Artifacts for Windows and macOS will be available automatically.

Heads up: there are 2 known crashes on shutdown on Linux that we hope to track down & fix before full OBS 27.2 release.

**Reply by WizardCM** [reply]
Smaller update: we've solved at least 2 crashes on shutdown related to CEF on Linux.

There does seem to be one more Linux crash that we've had less luck fixing, but it has relatively strict reproduction steps so we're hoping it won't hit too many users during beta. We hope to revisit the issue either during beta or very soon after release.

Linux crash (on shutdown) reproduction steps for anyone who'd like to try their hand at debugging:

1. Add a Custom Browser Dock of the Twitch homepage
2. Ensure a video stream is playing in the dock
3. Close OBS

**Comment by TheRedM** [comment]
Is there any plans for browser source on arm linux? I'm actually trying to make it work using Spotify pre built CEF binary, this is my first time digging in a project, i hope i can make it work.

**Reply by tytan652** [reply]
Since aarch64(arm64) is planned for macOS and Windows, Linux could join it too. But for arm (32bit) I don't know.
Note: from 27.2, the CEF for Linux include custom patches.

https://github.com/obsproject/cef/tree/MediaHandler

**Reply by TheRedM** [reply]
I just successfully built it for linux arm64, working good so far.
Funny thing i had never checked this branch, i just went straight to Spotify pre built CEF and did some digging (bazukas old obs-linuxbrowser) and it worked.

**Comment by amazingfate** [comment]
The cef binaried are uploaded to Fastly, but the source code is at: https://github.com/obsproject/cef. Is there an instruction to let people know how to build the binaries from source code? I'm trying to build one for linux arm64.

**Reply by derrod** [reply]
I don't believe we do anything special, just following Google's original compilation instructions should get you there.

That said we have a prebuilt arm linux version available: https://cdn-fastly.obsproject.com/downloads/cef_binary_5060_linux_aarch64.tar.xz

**Reply by amazingfate** [reply]
Thanks! I didn't find it in the docs. I will use it for my arm64 obs.

**Reply by Vitalicus** [reply]
what are the commands to compile?

**Reply by amarmashruwala** [reply]
Hi All, how do i install OBS-studio on Armbian jammy desktop ? thanks

**Reply by amazingfate** [reply]
I packaged obs v29 in my ppa: https://launchpad.net/~liujianfeng1994/+archive/ubuntu/rockchip-multimedia

**Reply by Fenrirthviti** [reply]
This discussion post is not an appropriate place to ask for support. Please use the Forums or Discord instead.

**Comment by amarmashruwala** [comment]
Ok, can you please share the link to the forum and discord sever?
Thanks

**Comment by WillFM** [comment]
Are there any plans for CEF updates (was hoping to see one with OBS 30)? I see the last update was from 2022, almost two years ago. OBS Browser is on CEF version based on Chromium 103, yet the latest version of CEF is now on Chromium 119. there are a lot of vulnerability fixes, performance improvements, and features added over the past two years.

**Reply by Fenrirthviti** [reply]
We cannot update, as the features we rely on are no longer available in newer releases.

**Reply by WillFM** [reply]
I didn't read the earlier posts thoroughly enough. I see mention of hardware acceleration, but no clear-cut mention of what features were dropped in CEF that OBS relies on.

Some googling, I'm seeing mentions about Offscreen rendering with hardware acceleration. Is this the dropped feature?

**Reply by pkviet** [reply]
> I didn't read the earlier posts thoroughly enough. I see mention of hardware acceleration, but no clear-cut mention of what features were dropped in CEF that OBS relies on.
> 
> Some googling, I'm seeing mentions about Offscreen rendering with hardware acceleration. Is this the dropped feature?

Texture sharing is not possible anymore. We used a patch which can't be applied for versions later than 103. We've tried but there are too many stuffs to revert in chromium. What we need is the feature to be ported to skia.

**Reply by Fenrirthviti** [reply]
Effectively. The patches to support OSR are not compatible with the new renderer, and need to be re-developed from scratch. The tl;dr is that CEF as a project does not want to keep dragging patches along for the renderer, and Chromium doesn't have motivation to implement it themselves, so someone needs to do the work which is extremely complex and delicate. Anyone with the requisite skills/knowledge to do it is not available to work on it.

**Reply by WillFM** [reply]
Appreciate the info!

**Comment by WizardCM** [comment]
Status Update:

I bring great news! Thanks to fantastic work by @reitowo and others, we're on track to upgrade CEF in OBS to a modern version once again!

### We've successfully built & tested CEF 6478 (Chromium 126) with all the bells & whistles we require.

This does not mean we'll be shipping an OBS update with a new Chromium immediately, as we'd like to perform extensive testing across all platforms and architectures.

Technical details for those curious can be found at the below obs-browser PR, and I'll provide a summary of the story so far (& the next steps).

* https://github.com/obsproject/obs-browser/pull/434

For many years, we've been using an unofficial patch to CEF and adapting it where we could for each new Chromium release. Chromium 103 & 104 made dramatic changes to their rendering backend, removing the paths that the patches depended on. We continued reaching out to various parties to find someone willing to rewrite the patches with modern technologies, without much luck.

The reason the patch was unofficial in the first place is because Chromium itself didn't really handle the use case, and the CEF maintainers didn't want to have to maintain code that could fundamentally break every single major Chromium version, which is absolutely understandable.

Earlier this year, @reitowo [posted](https://bitbucket.org/chromiumembedded/cef/pull-requests/285/reimplement-shared-texture-support-for-viz#comment-464178305) in the abandoned CEF PR to check if any progress had been made, and as there was no response, they got to work. Less than two weeks later, they [posted](https://bitbucket.org/chromiumembedded/cef/pull-requests/285/reimplement-shared-texture-support-for-viz#comment-468095494) an initial implementation, and since then we've coordinated as best we could to get all the necessary bits in place, eventually resulting in **eight** Chromium PRs which were successfully merged throughout February and March.

As the Chromium maintainers coordinated to get the API fleshed out and the changes merged, we began work updating our own code in preparation. Eventually, in April, the changes to CEF were [merged](https://bitbucket.org/chromiumembedded/cef/pull-requests/734), meaning new official CEF builds going forward would provide what we needed.

This meant it was time to start in-depth testing, and it's where we hit a minor snag. Crashes on shutdown when "Control audio via OBS" was enabled - which @pkviet promptly tracked down and [fixed](https://bitbucket.org/chromiumembedded/cef/pull-requests/757/fix-crash-on-exit-due-to-audio-capturer) in May. With all known side effects taken care of, it was time to make our first non-Windows builds, as building CEF takes a while & there's no point repeatedly building something you can't yet use. CEF 126 builds were created last week for all 3 platforms, and I went ahead and performed the tests across Windows, macOS & Linux, verified everything works, and pushed the changes to the PR.

### So, next steps?

1. We're currently in OBS 30.2 beta testing, therefore **no major changes will be added** until after OBS 30.2's release, so in the meantime..
2. I will await a second opinion on the code in the above linked obs-browser PR, which **will be merged post-OBS 30.2 release**
3. Chrome 126 reached [Stable channel](https://chromestatus.com/roadmap) **5 days ago**, so in a couple of weeks CEF will start providing [official stable builds](https://cef-builds.spotifycdn.com/index.html) to match
4. When that occurs, we'll make final OBS CEF 126 packages that are ready to be shipped & tested for all platforms. That's when I'll post an update below
5. If testing goes well, a future OBS version later this year or very early next year will be updated with the latest stable CEF at time of OBS beta, which could even be CEF 127 or 128 depending on timelines (and testing results).

Thanks to everyone for your patience thus far. I'm very excited about being able to bump CEF more frequently going forward & providing improved tools for web overlay developers.

**Reply by WizardCM** [reply]
As promised, the first follow-up update!

We now have **builds available for testing** running Chromium 127! You can find the artifacts [here](https://github.com/obsproject/obs-studio/actions/runs/10607923247?pr=11213#summary-29401392327) (and Flatpak [here](https://github.com/obsproject/obs-studio/actions/runs/10612662766?pr=11218#summary-29414815683)). :tada:

* https://github.com/obsproject/obs-studio/pull/11213
* https://github.com/obsproject/obs-studio/pull/11218
* _**Note:** There is a known crash-on-shutdown on Linux. A fix is being reviewed, and will be merged before release._
  * https://github.com/obsproject/obs-browser/pull/450

You'll notice this build is 127, not 126 like previously tested. In the time it took us to get everything going, 127 went stable. That's good news - it means the Chromium and CEF changes we were waiting on continue to be functional and supported in subsequent CEF releases!

If you're looking to test these builds, please be sure to provide your feedback on the PR - both good and bad.

If you're a web developer, you can find the full list of changed web platform features [here](https://chromestatus.com/newfeatures?q=browsers.chrome.desktop%253E103%2520browsers.chrome.desktop%253C%253D127).

---- 

Once this upgrade is released to users globally, I'll write a separate post about the challenges in upgrading to Chromium 128+ due to CEF exclusively providing the Chrome Runtime going forward. It's not impossible to upgrade, but contains breaking changes (and lack of user migration path) that we have to handle gracefully.

**Comment by WizardCM** [comment]
Status Update:

As promised in August, **OBS 31 shipped in December with Chromium 127**, and while we've hit a couple snags around [performance on Linux](https://github.com/obsproject/obs-studio/issues/11485), [stuttering animations](https://github.com/obsproject/obs-browser/issues/468), [controller overlays no longer working](https://github.com/obsproject/obs-browser/pull/471), and some crashes we're still tracking down (please submit crash reports!), overall the upgrade has been a success.

Our next focus, as mentioned previously, is working towards supporting the Chrome Runtime, which is required to allow us to upgrade past Chromium 128.

It's a doozy.

Long story short: we have to build multiple additional protections and a user migration path for browser profile data, and it'll take a while. We hope to have this fully done sometime this year, but due to many potential unknowns, I don't want to provide a concrete timeline just yet. I'll provide updates as we make progress.

---

If you're curious about the work involved, here goes.

Since its inception, CEF3 has used what's called the "Alloy" runtime. This is a custom layer on top of the Chromium core that has, over the years, grown to expose a variety of Chromium functions in a relatively stable, extensible way. However, this meant that those developing CEF had to manually implement everything - if Chromium added a new feature or API, it wouldn't be available to users of CEF until someone wrote an implementation, it was submitted, reviewed, and merged. This practically meant that the CEF developers were essentially attempting to re-implement most of the Chromium API with their own twist. Ultimately, this proved unsustainable.

So, a couple of years ago, work began on providing a more direct path between Chromium and the users of CEF (ie. us) - the Chrome Runtime. This new system would run the Chromium frontend, where a lot of user-facing features of Chromium reside, so that the CEF developers would no longer have to rewrite certain Chromium frontend functionality into CEF. In CEF 128, the Chromium runtime was made default and the Alloy runtime was entirely removed.

While this is great in theory, and ultimately fantastic for CEF's future, it puts us in a bit of a bind for the short term. A big upside of the Alloy runtime is in its limitations - functionality is only exposed to users if the application, OBS, chooses to provide an interface. With the Chrome Runtime, all of Chromium is now magically accessible. That means Chrome Extensions, the password manager, Cast.., popout picture-in-picture, Live Captions, Set as Default Browser, User Profiles, Incognito Mode, Printing, and most problematic of all, a method of opening a standard Chromium window owned by the OBS executable but not controlled by it. Lots of avenues for users to cause trouble, and more ways for Chromium to wreak havoc in the stability of OBS.

Additionally, certain features we like/prefer in the Alloy runtime either aren't available in the Chrome Runtime, or don't work the same way. One of these is User Profiles. Browser sources and Custom Browser Docks share cookies, storage, and events, allowing a browser overlay to be controlled by a dock, and only requiring a single login. Service integration, like Twitch, YouTube, and Restream, use their own separate storage to avoid cross-contamination and to improve security. Unfortunately, both of these paths have changed dramatically in the Chrome runtime, to the extent that service integration (by default) uses incognito mode, so it loses its information on every OBS restart. That's problematic.

To maintain a sane scope for the first iteration, we're planning on essentially recreating the Alloy runtime experience while using the Chrome Runtime. So, at a bare minimum, we're looking at

1. Disabling various functionality exposed by Chrome that we don't want or need. Some of these may be re-enabled in future based on testing and if we believe there are valid use cases for it
2. Performing an automatic one-time migration of browser data so that users don't have to login & configure everything again
3. Fixing service integration
4. Blocking various behaviours that don't make sense for us, such as marking "OBS Studio" as the default web browser in your operating system
5. Changing various default Chrome settings, such as disabling autofill, autocomplete, search suggestions and more

We'll also need to test more than ever before to ensure our users get the best experience. I'd rather we delay than release something partially functional. For all I know, there are additional things we may need to request CEF to change, or areas I haven't touched yet that might be completely broken. My current testing is being performed with Chromium 133, but that will continue to rise as time passes.

---

Thanks again for everyone's patience while we continue our efforts to provide a stable, up-to-date browser experience within OBS.

**Reply by imran179971** [reply]
can you give me details about this

**Reply by hoshinolina** [reply]
We're planning to ship up-to-date CEF (137 at the time of this writing, 138 in a couple days when I update it) with OBS 31.1.0 in the Fedora RPM package, when it is released. I previously packaged the obs-cef fork for Fedora in F40, but making old versions of CEF build and continue to build with up to date compilers and system libraries was impractical, so we had to drop it for F41. Since OBS now works with upstream CEF and CEF has a stable API/ABI from 133 onwards, this makes it much, much easier to support the Browser Source going forward and keep the browser engine up to date (with security patches, etc.) while compiling everything from source and complying with all the Fedora packaging requirements (our CEF build is based on and tracks Chromium releases for Fedora, so all the OS/distro/platform-specific stuff should be taken care of as well as it is for the Chromium browser, which gets a lot of testing and usage). CEF is packaged as a system-wide package so it can be updated independently from OBS (and we handle all the API/ABI versioning stuff to match the upstream CEF guarantees). Currently the assumption is that OBS works with CEF API version 13304 (since you're testing with CEF 133, I think that makes sense and seems to work as well as expected).

I know this isn't officially supported by upstream OBS yet, but I think it makes sense to ship it now since the alternative is no browser support at all in the package. Fedora users are supposed to file bugs in our downstream bug tracker, so please feel free to direct any users that wind up here to https://bugzilla.redhat.com/. That said, I hope this also helps OBS by providing testing for the Chrome Runtime change. If there is anything I can do to help direct useful and wanted feedback back upstream, please let me know! I'm also going to be using this all personally in my streaming soon.

**Reply by HurricanePootis** [reply]
I'm trying to get something similar done for [Arch Linux's](https://gitlab.archlinux.org/archlinux/packaging/packages/obs-studio/-/merge_requests/2) OBS build. I really appreciate the work you did for Fedora's.

**Comment by imran179971** [comment]
what is this bro can anybody explain me+923287003040 it's my WhatsApp number

**Comment by WordlessEcho** [comment]
Hi there, I’m looking forward to use a feature in Chrome 140+. Any updates you can share with us?

**Comment by WizardCM** [comment]
Status Update:

Apologies for the extensive delay in this one. I didn't want to post something without concrete movement forward. I wasn't kidding about this one being a doozy - and burnout got me good. But, here we go..

I have _finally_ opened a pull request to the OBS Browser repository with support for CEF 128 (Chrome 128) up to CEF 7727 (Chrome 147) alongside the Chrome Runtime.

* https://github.com/obsproject/obs-browser/pull/523

You can read the full changes and reasoning behind various decisions in the PR description. Everthing I mentioned in my previous status update applied, especially the difficulties involved - feel free to read that post for full context.

The changes I've submitted are now awaiting review. I'm hoping that can happen in time for the next major OBS update (33.0), which will occur after the next scheduled release 32.2. That is up to availability of OBS team members, and testing performed by the wider community.

Additionally, due to how little my code has changed during my testing through versions 128 to 147 (including 132, 136, 143 and 145) it seems I was correct that we'd be able to update more consistently going forward. My current thinking is that we should ideally target updating during major OBS versions (33.0, 34.0, etc). Per [Warchamp's article in 2024](https://obsproject.com/blog/obs-studio-versioning-changes), that should be once a year, with obvious exceptions when a security patch is required. Personally, I think this would strike a good balance between staying up to date and avoiding a large download every 4 months.

Thank you all for your patience. Looking forward to posting a new update once the PR is merged, which will likely result in an update to [Chrome 15x](https://chromestatus.com/roadmap).

**Reply by WizardCM** [reply]
As promised, the first follow-up update!

We now have **builds available for testing** running Chromium 150! You can find the artifacts [here](https://github.com/obsproject/obs-studio/actions/runs/35032216796?pr=13900#artifacts) for all platforms, including Flatpak. :tada:

* https://github.com/obsproject/obs-studio/pull/13900

The build we're planning to upgrade to is specifically 150, as it is the newest stable version we have confirmed working, and have performed extended testing with. A quick test of 151 occurred, however it crashed within Chromium code. 152 seems to fix it, but I'd rather go with a more extensively tested version.

If you're looking to test these builds, please be sure to provide your feedback on the PR - both good and bad.

If you're a web developer, you can find the full list of changed web platform features [here](https://chromestatus.com/features?q=browsers.chrome.desktop%253A128..150).

---- 

A reminder - this upgrade contains migration code for website data and storage. I have attempted to cover all important data in the migration, and occurs one-time. Information about migration is printed to the log to help troubleshooting. If a user downgrades OBS versions, or we end up reverting the Chromium upgrade, the user's old browser profile will be used instead.
